### PR TITLE
[ML] Full cluster restart tests for migration

### PR DESCRIPTION
**Pull request by davidkyle**
Similar to the upgrade tests in #36425

Creates jobs in the old cluster opening one of them, then in the upgraded cluster asserts that the closed jobs have been migrated and can be opened. Asserts that jobs left open during upgrade get an assignment

### CODE REVIEW

**Comment by davidkyle** [comment]
run gradle build tests 2

File: x-pack/qa/full-cluster-restart/src/test/java/org/elasticsearch/xpack/restart/MlMigrationFullClusterRestartIT.java [hunk 1]
```diff
@@ -0,0 +1,258 @@
...
+
+                Map<String, Object> stats = entityAsMap(response);
+                List<Map<String, Object>> jobStats =
+                        (List<Map<String, Object>>) XContentMapValues.extractValue("jobs", stats);
+
+                assertEquals(jobId, XContentMapValues.extractValue("job_id", jobStats.get(0)));
+                assertEquals("opened", XContentMapValues.extractValue("state", jobStats.get(0)));
+                assertThat((String)XContentMapValues.extractValue("assignment_explanation", jobStats.get(0)), isEmptyOrNullString());
+                assertNotNull(XContentMapValues.extractValue("node", jobStats.get(0)));
+            } catch (IOException e) {
```
Comment (droberts195) [inline x-pack/qa/full-cluster-restart/src/test/java/org/elasticsearch/xpack/restart/MlMigrationFullClusterRestartIT.java]: Why do we need to ignore `IOException`s here?

Comment (davidkyle) [reply x-pack/qa/full-cluster-restart/src/test/java/org/elasticsearch/xpack/restart/MlMigrationFullClusterRestartIT.java]: Well the assert busy will retry and eventually timeout, either way the test will fail but I admit the cause will be lost

Comment (davidkyle) [reply x-pack/qa/full-cluster-restart/src/test/java/org/elasticsearch/xpack/restart/MlMigrationFullClusterRestartIT.java]: I don't do it elsewhere I'll clean that up

Comment (droberts195) [reply x-pack/qa/full-cluster-restart/src/test/java/org/elasticsearch/xpack/restart/MlMigrationFullClusterRestartIT.java]: I think an exception other than an `AssertionError` will stop the busy loop.  So if an `IOException` can be thrown that _shouldn't_ terminate the busy loop then it needs to be caught here.  But maybe there isn't.

Comment (davidkyle) [reply x-pack/qa/full-cluster-restart/src/test/java/org/elasticsearch/xpack/restart/MlMigrationFullClusterRestartIT.java]: Yes assert busy only catches  `AssertionError` and the IOException should fail the test and not be caught here. I'm gonna blame intellij and remove the try...catch

**Review by droberts195 — APPROVED** [review]
LGTM

**Pull request merged**